src/ctl/src/cmd_impl/await_tree/analyze.rs

```diff
@@ -0,0 +1,185 @@
+    ///
+    /// Each Tree here maps to a specific actor and fragment of a streaming graph.
+    /// The diagram below shows the logical structure of a stuck streaming graph:
+    /// -------------------------------------------------------------------------------
+    ///    OB(Source Executor)  ->  OB  ->  BN  ->  IB  ->  IB(Materialize Executor)
+    /// -------------------------------------------------------------------------------
+    ///   |   waiting for new epcoh   | bottleneck |      Waiting for input         |
+    /// -------------------------------------------------------------------------------
+    ///
+    /// A typical look of OB Tree/IB Tree:
+    /// ```text
+    /// Actor 123456: `XXXXX` [1595.653s]
+    ///   Epoch 7509626714259456 [!!! 1582.513s]
+    ///     CdcFilter 1DFBC0000271D [!!! 1582.423s]
+    ///       Merge 1DFBC00000000 [!!! 1582.423s]
+    ///         LocalInput (actor 122807) [!!! 1582.423s]
+    /// ```
+    /// Note:
+    /// For an OB Tree, all the executors in the tree are waiting for source.
+    /// However, the source executor is also blocked by barrier collection.
+    /// For an IB Tree, all the executors in the tree are waiting for the input data from
+    /// the bottleneck executor, which is the upstream of the IB Tree.
+    ///
+    /// A typical look of BN Tree:
+    /// ```text
+    /// Actor 123456: `XXXXX` [1595.673s]
+    ///   Epoch 7509625856917504 [!!! 1590.993s]
+    ///     Materialize 9E2000000000D [!!! 1590.993s]
+    ///       Project 9E2000000000C [!!! 1590.993s]
+    ///         Project 9E2000000000B [!!! 1590.993s]
+    ///           Project 9E2000000000A [!!! 1590.993s]
+    ///             HashAgg 9E20000000009 [!!! 1590.993s] <== Bottleneck
+    ///               Merge 9E20000000008 [980.020ms]
+    ///                 LocalInput (actor 647685) [980.020ms]
+    /// ```
+    /// Note:
+    /// There is usually a bottleneck executor throttling the whole graph. So we can
+    /// detect the bottleneck by checking the elapsed time of the bottleneck executor
+    /// and the average elapsed time of its children. If the elapsed time of the
+    /// bottleneck executor is much larger than (eg, 5x) the average elapsed time of its
+    /// children, we can say that the bottleneck executor is the bottleneck of the graph.
+    ///
+    /// A special look of IB Tree:
+    /// ```text
+    /// Actor 123456: `XXXXX` [1595.653s]
+    ///   Epoch 7509625856917504 [!!! 1591.003s]
+    ///     Union 9E20200000007 [1.000s]
+    ///       Merge 9E20200000003 [1.000s]
+    ///         LocalInput (actor 647689) [1.000s]
+    /// ```
+    /// Note:
+    /// In this case, although the bottleneck actor(BN Tree) throttles the whole graph,
+    /// the bottleneck actor is still yielding output to downstream actors. A typical
+    /// case is JOIN amplification. So the corresponding actors are actively processing
+    /// the data but the EPOCH span is blocked.
+    pub fn is_bottleneck(&self) -> bool {
```

Comment:
Hmmm, the algorithm looks much easier than I thought initially. So there's no need to traverse between actors (based on the `Input` and `Output` spans) at all?

Reply:
To be honest, I'm not entirely sure this algorithm works in every situation. However, in most cases where there's only one bottleneck actor, it should perform as expected. I can think of a few edge cases:

1. If the bottleneck's span doesn't call `instrument_await`, the bottleneck's tree might not exhibit the typical "slow parent with fast children" signal.
2. If an MV or the entire graph has multiple fragments acting as bottlenecks, I'm uncertain what the outcome would be.

At present, this tool doesn't completely eliminate manual effort. It definitely requires ongoing optimization and the addition of new rules.

Reply:
> 2\. I'm uncertain what the outcome would be.

Shall we output all of them, ordering by some sort of confidence or probablity?

Reply:
Tend to leave it for future. At present, we don't have too much experience dealing with these edge cases. We can add some new rules when we can summarize some tips for them. For now, if the current `bottleneck_detect` return nothing, the on-caller should just use `transcribe` to convert the json await tree to text and mannually analyse the text await tree.